### PR TITLE
Implementing proposed api of #3

### PR DESCRIPTION
**Pull request by Fuzzyma**
So now we have 2 PRs. Thats non optimal but we will somehow manage it.
As discussed in #3 I proposed the following api:

>- `zoom(level, point)` - when called without an argument it returns the zoom (-level)
>- `animate().zoom(level, point)`
>- `zoom(level)` should use the center of the canvas as focus point as default
>- panZoom caculated a zoom lvl based on its factor and uses `zoom` to do the zoom - that could be a bit of overhead, though. Or it zooms just directly in lvls... dunno

This is now implemented in this branch.
Here is an example on how to use it:

```js
var canvas = SVG('id').size(1500, 750).viewbox(1500,750)

// something we can see on the canvas
canvas.rect(1000, 500).move(250, 125)

// zoom in
canvas.zoom(2)

// reset zoom
canvas.zoom(1, new SVG.Point()) // defaults to 0,0

// animate zoom out
canvas.animate(1000, '<>').zoom(0.5)

// get current zoom level
var zoom = canvas.zoom()

// activate panZoom
canvas.panZoom()
```

The greatest pitfall of this impementation is, that you cant specify a focus point when animating the zoom. For this a few changes in the fx module are required which I want to discuss with @rmittro before. For now it always takes the default point which is the center of the canvas.

### CODE REVIEW

**Comment by dotnetCarpenter** [comment]
On the surface, this looks good. But I just tried on Android Browser 4.1, which is the oldest browser I need to support and neither zoom or pan worked. I'm debugging it right now. I suspect it has to do with prefixes.

**Comment by Fuzzyma** [comment]
Well - I did not intend to add workarounds for all this old browsers out there. 

I would suggest that the user adds polyfills for the needed features but that's nothing the plugin should care about. It does everything the "standard" way :)

**Comment by dotnetCarpenter** [comment]
Yes you are right. I only meant to say, that I don't think it's anything serious. I'm just debugging, which is kinda hard on Android Browser.

**Comment by Fuzzyma** [comment]
Oh yes I feel you!

**Comment by dotnetCarpenter** [comment]
I think I found it. You're using the [exponentiation operator](https://github.com/svgdotjs/svg.panzoom.js/blob/baa7c23e5982fcc36acb322739f0a2ba50546b5a/svg.panzoom.js#L56) which IFAIK not a standard until ES2017. I think we have to use `Math.pow`.

**Review by dotnetCarpenter — COMMENTED** [review]
I think I found it. You're using the [exponentiation operator](https://github.com/svgdotjs/svg.panzoom.js/blob/baa7c23e5982fcc36acb322739f0a2ba50546b5a/svg.panzoom.js#L56) which IFAIK not a standard until ES2017. I think we have to use `Math.pow`.

**Comment by Fuzzyma** [comment]
Right - looks like that happend by accident :D. Feel free to fix!

**Comment by dotnetCarpenter** [comment]
OK, I should be able to get this running today. Now I don't get any errors, but panning is reversed and pinch zoom doesn't work at all. Did you get it to run on a touch device and if yes, which one? I'm gonna create some Android Virtual Devices to debug this.

**Comment by Fuzzyma** [comment]
Yes I tested it with my android 6 with chrome as browser. The cyanogenmod standard browser wasn't able to handle it at all. 

In chrome it was working as everyone would expect

**Comment by dotnetCarpenter** [comment]
So that is Android Marshmallow, API level 23, right?

**Comment by Fuzzyma** [comment]
No clue about the api level but marshmallow yes!

**Comment by dotnetCarpenter** [comment]
I want the zoom event to also fire when pinch zooming. I think to either make `pinchZoom` use `zoom` or fire the `"zoom"` event in `pinchZoom`.

Why do we need [`focusP`](https://github.com/svgdotjs/svg.panzoom.js/pull/4/files#diff-3cbec73df56a454d7dd0e325f4cb6539R73) in `pinchZoom`?

File: svg.panzoom.js
Comment (dotnetCarpenter) [inline svg.panzoom.js]: This looks like a neat trick but I'm not sure why you do it. Can you elaborate?

File: svg.panzoom.js
Comment (dotnetCarpenter) [inline svg.panzoom.js]: I'm getting **Uncaught TypeError: Cannot read property 'clientX' of undefined** here on Android 4.1, sometimes. I wonder if it has to do with the option object in `SVG.on` becoming a capture boolean when options are not supported and then we're are missing some event bubbling.

But I have been debugging this a lot now and it's not 100% consistent with that. If it the case then it should also happen in modern browsers when the capture flag is set to `true`, but it does not.

**Review by dotnetCarpenter — APPROVED** [review]
I'm not sure about the `zoomAmount` logic - the fact that it makes the zoom smaller the more you are _zoomed in_ but it's not a big deal for me.

With the exception of the Android Browser 4.1 bug that _sometimes_ appears, I'm quite happy with the code. Good work! :+1: 

PS. I added the zoom event trigger to `pinchZoom` as well - all zooming should be detectable through that single event.

File: svg.panzoom.js
Comment (Fuzzyma) [reply svg.panzoom.js]: In svg.js we lately introduced the `event()` method which returns the last fired event of that element. By firing an event and immediately checking the getting that event, we could check the `defaultPrevented` flag. It is true, if any event handler called `preventDefault()` on the event.

So in case it *is* true, we return because our behavior should not get executed

File: svg.panzoom.js
Comment (Fuzzyma) [reply svg.panzoom.js]: Mh we might check if lastTouches[0] is always defined at this point to handle this error.
But generally there should *always* be a last touch because `pinchZoomStart` takes care of that... weired

**Comment by Fuzzyma** [comment]
The focusPoint is the point between your fingers in which you zoom into. It can change when you pan while zooming. Thats why it becomes a bit tricky to calculate the corresponding matrix.
You translate the box to its origin, scale it and translate it back to the new focus point.

Beside this I have a new idea: I saw on many websites that use inline googlemaps, that they do not pan on touch drag. You have to use 2 fingers instead so it does not interefer with the normal scroll on the phone.
I like this idea a lot and would like to propose it. What do you think about that? We might consider it as an default behavior while hiding the touchdrag behind an option flag or removing it alltogether

**Comment by dotnetCarpenter** [comment]
>Beside this I have a new idea: I saw on many websites that use inline googlemaps, that they do not pan on touch drag. You have to use 2 fingers instead so it does not interefer with the normal scroll on the phone.
I like this idea a lot and would like to propose it. What do you think about that? We might consider it as an default behavior while hiding the touchdrag behind an option flag or removing it alltogether

I think that is a great idea. If it is possible to implement the current `touchdrag` outside of the plugin, then I think that is the route to take. We don't have to implement every possible configuration but it is still possible to roll out your own implementation in _user-land_. Don't know if that is possible though.

File: svg.panzoom.js
Comment (dotnetCarpenter) [inline svg.panzoom.js]: @Fuzzyma not sure if it is currently possible to set the point, when animating a zoom. You mention that in the end of the [first comment](https://github.com/svgdotjs/svg.panzoom.js/pull/4#issue-224416418).

Comment (Fuzzyma) [reply svg.panzoom.js]: this feature was already pushed and will be available in the next release. It doenst hurt to include it here already. The old version just does nothing with the second parameter

File: svg.panzoom.js
Comment (Fuzzyma) [reply svg.panzoom.js]: @dotnetCarpenter about the capture-flag: I found this site: https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md

It also explains how to polyfill the feature or work around it

**Comment by dotnetCarpenter** [comment]
In order to implement two finger pan, that implies that you can only pan if you pinch zoom, right?

We detect pinch zoom by counting touch events, if there is two, then we use the distance between these two to set the zoom level.

So while we don't get in conflict with the page scroll, we come in conflict with our own pinch zoom. Or am I missing something?

**Comment by Fuzzyma** [comment]
That's the thing I explained with the focus point. You can pan while you zoom already. So the implementation is already there. We just need to remove the touch drag

**Comment by dotnetCarpenter** [comment]
If it is possible to re-implement single-touch pan in user-land, then I'm all for it.

Let me see if we're missing something..

**Review by dotnetCarpenter — CHANGES_REQUESTED** [review]
If a user want to implement another kind of input, they will need the [`zoomInProgress`](https://github.com/svgdotjs/svg.panzoom.js/pull/4/files#diff-3cbec73df56a454d7dd0e325f4cb6539R100) property. Like we do in `panStart`.
So I think that should be made public.

I don't see any other show stoppers. We will save 7 lines of code, will get simpler event handling logic and a nicer touch experience.
*BUT*, we are putting a huge burden on developers that want single-touch pan. But let's wait and see if anyone will ever request that. In the mean time, I pulled out some code to get a working implementation in user-land. I can post it on the wiki, if people *really* want it.

File: svg.panzoom.js
Comment (dotnetCarpenter) [inline svg.panzoom.js]: After reading [the event listener options you gave me](https://github.com/WICG/EventListenerOptions/blob/gh-pages/explainer.md), I worry about the `document` event listeners. I think we only need to listen on the `panZoom` element and not the entire document. Also, what will happen if I have 2 small maps on a page, using svg.panzoom.js?

**Comment by Fuzzyma** [comment]
I think we shoulnt make a property public but instead fire zoomStart and zoomStop events. The user then can track it on its own.

File: svg.panzoom.js
Comment (dotnetCarpenter) [reply svg.panzoom.js]: Our `SVG.on` implementation is: `node.addEventListener(ev, l, options || false)`

If we pass `{ passive: false }` in a browser that does not support options, then it translate to `node.addEventListener(ev, l, true)`.

Perhaps we could do
`node.addEventListener(ev, l, (options && options.passive && options) || false)`

That would translate to false, when options is `undefined` or `{ passive: false }` or `false`. And to `{option: true}` or `true` depending on browser support.

**Comment by dotnetCarpenter** [comment]
Conceptually, you're right. But events looks very expensive, every time I profile this. And conceptually perhaps, harder to understand for a user, than just having a read only property. Hmmm... not sure.

Well, we have to pick one. I think we just go with events, that's how most things are communicated.

File: svg.panzoom.js
Comment (Fuzzyma) [reply svg.panzoom.js]: As you can see the document listeners are only for touchmove and touchend. Not for touchstart.
This is generally needed to make sure, that you also can continue panning when you out of the borders of the orginally box. Restricting the users pan to the box only feels clunky and is not what the user would expect.

So thats why we have this. The handlers on document will never interfer with a second map because they are bound on touchdown so we are sure the user selected an item to pan

Comment (dotnetCarpenter) [reply svg.panzoom.js]: ahh - clever

**Comment by Fuzzyma** [comment]
That's interesting. What exactly is expensive? Just firing events? Or the case that listener are bound to the event already?

**Comment by Fuzzyma** [comment]
Just realized that we need the normal pan for mouseinput. So the code stays. But it becomes a bit less confusing at least

**Comment by dotnetCarpenter** [comment]
I regret only have so little time for this. Every time I get back, I have to familiarize myself with the code and the design decisions and discussion. Now I really want to land this. I have a use-case where I need to pan-zoom a world map, display a mini map that can only pan (reflecting pan on the big world-map) and display  a title for each country on the map. The latter doesn't really have anything to do with svg.panzoom.js but it's a good use-case.

I will hack-a-way and make changes where needed. If we can land this the next few days - I'm a happy man :)

**Comment by dotnetCarpenter** [comment]
@Fuzzyma about firing events. It seems that Firefox, does a lot of string stuff - just when firing events. I will re-profile and post the result. It might be nothing, just something I've noticed.

**Comment by Fuzzyma** [comment]
Oh man, why do I never push my changes? I made thsi changes, too. Will merge when I am at home. I had no chance getting this finished yesterday

// EDIT: Is it okay for you to wait with further pushes so that I can merge my stuff in?

**Comment by dotnetCarpenter** [comment]
Sounds good. Let me know what you think. In the mean time I will implement `zoomStart`/`zoomEnd` and `panStart`/`panEnd`.

File: svg.panzoom.js
Comment (dotnetCarpenter) [inline svg.panzoom.js]: The current `fire('zoom'` implementation is problematic in regard to `zoomStart`/`zoomEnd`. I'll need to refactor line 159-162 and possibly a lot more, before it makes sense.

Comment (dotnetCarpenter) [reply svg.panzoom.js]: Another thing. Why do we `return this.viewbox(box)`? We don't seem to use it for anything. Animation work fine without and `wheelZoom` does not return the value from `zoom`.

**Comment by Fuzzyma** [comment]
my work is pushed. Now do as you widh :).
I migrated to gulp. So run `npm i && gulp` for building.

**Comment by dotnetCarpenter** [comment]
@Fuzzyma thanks! Can you tell me why we're are `return this.viewbox(box)`? It seems that we're not using it for anything. Animation works fine without and `wheelZoom` does not return the value from `zoom`.

File: svg.panzoom.js
Comment (Fuzzyma) [reply svg.panzoom.js]: zoomStart and zoomEvent shoould not reside in the zoom function. It should be in touchstart/end imo. 

`return this.viewbox(box)` actually returns `this` so I saved a few bytes (chaining ftw)

Comment (dotnetCarpenter) [reply svg.panzoom.js]: That does not cover `wheelZoom` and programmable zoom. But I think we can get away with only having one `zoom` event. I'll need to play with it a little

**Comment by Fuzzyma** [comment]
Just don't overcomplicate things :). Maybe we can merge this first and
create a new pr for the new stuff?

Am 06.06.2017 19:13 schrieb "Jon Ege Ronnenberg" <notifications@github.com>:

**Comment by dotnetCarpenter** [comment]
Hmm.. I don't know if it's complicating things but a `panStart` and `panEnd` makes sense to have so that you can change appearance on the website. For completeness `pinchZoomEnd` also makes sense since we now have `pinchZoomStart`.

I want to change `[].slice.call(ev.touches)` to `ev.touches` since we don't need `ev.touches` to be an Array but any object that has index. We use `[0]` to access values. Is there any reason I'm not aware of that merit converting `ev.touches` to an Array?

**Review by dotnetCarpenter — APPROVED** [review]

**Comment by dotnetCarpenter** [comment]
@Fuzzyma One last thing. We should document `SVG.Point`. Especially because I reference it in the event documentation in the README

**Comment by Fuzzyma** [comment]
Yes SVG.Point should be documented. However, in the case of this plugin it's worth noting that the user don't really need svg.point. It's completely valid to pass an object with x and y property instead. So that's something you might want to include, too.

**Comment by Fuzzyma** [comment]
To the ev.touches thing: I think the slice is not really needed. It's there for integrity so you have a real array in every case (so you can filter, map... it) 

Btw is there a reason why you committed the package lock.json? I think this file does not belong into the repository

**Comment by dotnetCarpenter** [comment]
Yes, you're right. I'll update the README and merge this in, if's ok with you?

**Comment by Fuzzyma** [comment]
Please also read my other post before merging :)

**Comment by dotnetCarpenter** [comment]
`package lock.json` is there to be sure we have the same dependencies installed across systems. Similar to how `yarn` is meant to be used. You can look it up - the consensus is the always include it in version control, though in this project, it's hardly an issue that we don't have the exact same version of gulp and its dependencies.

Let's introduce `[].slice.call()` when we needed it and not before.

Is there anything else?

**Comment by Fuzzyma** [comment]
Imo lock files are great for projects but not for plugins and libraries. We gave our dependencies well defined in the package.json and we want allow the user to update svg.js if necessary. With a lock file you basically force him to one version. 
Correct me if I am wrong. I was just a bit upset about this HUGE file 

Nothing else from my side

**Comment by dotnetCarpenter** [comment]
Good point! removed...

I'll also squash the commits before merge....

**Comment by dotnetCarpenter** [comment]
manually merged

**Pull request closed without merging**
